compiler/rustc_pattern_analysis/src/lib.rs

```diff
@@ -21,7 +21,45 @@ rustc_fluent_macro::fluent_messages! { "../messages.ftl" }
 
 use std::fmt;
 
-use rustc_index::Idx;
+#[cfg(feature = "rustc")]
+pub mod index {
+    // Faster version when the indices of variants are `0..variants.len()`.
+    pub use rustc_index::bit_set::BitSet as IdxSet;
+    pub use rustc_index::Idx;
+    pub use rustc_index::IndexVec as IdxContainer;
+}
+#[cfg(not(feature = "rustc"))]
+pub mod index {
+    // Slower version when the indices of variants are something else.
+    pub trait Idx: Copy + PartialEq + Eq + std::hash::Hash {}
+    impl<T: Copy + PartialEq + Eq + std::hash::Hash> Idx for T {}
+
+    #[derive(Debug)]
+    pub struct IdxContainer<K, V>(pub rustc_hash::FxHashMap<K, V>);
```

Comment:
Can we use the rustc_data_structures types here instead and avoid the extra crate dep? (It's still the same under the hood, but makes it easier to tweak globally).

Reply:
> Can we use the rustc_data_structures types here instead [...]

`rustc_data_structures` is not compatible with rust-analyzer due to its use of nightly features. I guess @Nadrieril could add a `feature = "nightly"` to that crate, but it seems unnecessary.

> Are IDs elsewhere still contiguous in the common case, even if not starting at zero? 

I don't believe so. As Nadrieril said above, they're global interning IDs, so they're probably really sparse.

Reply:
No because rust_analyzer wants to avoid the `rustc_data_structures` dependency